relational-query/src/Database/Relational/Monad/BaseType.hs

```diff
@@ -52,44 +64,50 @@ askConfig =  qualify askQueryConfig
 
 -- | Unsafely type qualified subquery into record typed relation type.
-unsafeTypeRelation :: ConfigureQuery SubQuery -> Relation p r
+unsafeTypeRelation :: ReadPlaceholders p ConfigureQuery SubQuery -> Relation p r
 unsafeTypeRelation = SubQuery
 
 -- | Sub-query Qualify monad from relation.
-untypeRelation :: Relation p r -> ConfigureQuery SubQuery
-untypeRelation (SubQuery qsub) = qsub
+untypeRelation :: Relation p r -> Record PureOperand p -> ConfigureQuery SubQuery
+untypeRelation (SubQuery qsub) = runReadPlaceholders qsub
 
 -- | 'PersistableRecordWidth' of 'Relation' type.
-relationWidth :: Relation p r ->  PersistableRecordWidth r
-relationWidth rel =
-  unsafePersistableRecordWidth . width $ configureQuery (untypeRelation rel) defaultConfig
-  ---                               Assume that width is independent from Config structure
-
-unsafeCastPlaceHolder :: Relation a r -> Relation b r
-unsafeCastPlaceHolder (SubQuery qsub) = SubQuery qsub
-
--- | Simplify placeholder type applying left identity element.
-rightPh :: Relation ((), p) r -> Relation p r
-rightPh =  unsafeCastPlaceHolder
-
--- | Simplify placeholder type applying right identity element.
-leftPh :: Relation (p, ()) r -> Relation p r
-leftPh =  unsafeCastPlaceHolder
+relationWidth :: Relation p r -> Record PureOperand p -> PersistableRecordWidth r
+relationWidth rel phs =
+  unsafePersistableRecordWidth
+    . width
+    . (`configureQuery` defaultConfig) --- Assume that width is independent from Config structure
+    $ (untypeRelation rel phs)
 
 -- | Generate SQL string from 'Relation' with configuration.
-sqlFromRelationWith :: Relation p r -> Config -> StringSQL
-sqlFromRelationWith =  configureQuery . (showSQL <$>) . untypeRelation
+sqlFromRelationWith :: Relation p r -> Record PureOperand p -> Config -> SQLWithPlaceholderOffsets'
+sqlFromRelationWith r p c =
+  uncurry withPlaceholderOffsets
+    . (collectPlaceholderOffsets &&& showSQL)
+    . (`configureQuery` c)
+    $ untypeRelation r p
 
 -- | SQL string from 'Relation'.
-sqlFromRelation :: Relation p r -> StringSQL
-sqlFromRelation =  (`sqlFromRelationWith` defaultConfig)
+sqlFromRelation :: Relation p r -> Record PureOperand p -> SQLWithPlaceholderOffsets'
+sqlFromRelation r p =  sqlFromRelationWith r p defaultConfig
 
 -- | Dump internal structure tree.
-dump :: Relation p r -> String
-dump =  show . (`configureQuery` defaultConfig) . untypeRelation
-
-instance Show (Relation p r) where
-  show = showStringSQL . sqlFromRelation
+dump :: PersistableWidth p => Relation p r -> String
+dump =  show . (`configureQuery` defaultConfig) . (`untypeRelation` defaultPlaceholders)
+
+instance PersistableWidth p => Show (Relation p r) where
+  show = showStringSQL . detachPlaceholderOffsets . (`sqlFromRelation` defaultPlaceholders)
+
+defaultPlaceholders :: PersistableWidth t => Record PureOperand t
+defaultPlaceholders = pwPlaceholders persistableWidth
```

Comment:
I found this API should be "unsafe".
Because this enables us to make placeholder `Record`s even when building a `Relation () a`, which doesn't actually receive any placeholder parameters.